### ISSUE TITLE
Units that are also in scope

### ISSUE THREAD

**Issue by arve0** [josdejong/mathjs#860]
Hi!
I expected that the unit `m` for meter should **not** be overridden by scope.

Example in REPL:
```js
$ node bin/repl.js
> scope = {}
{}
> math.eval("m = 10kg", scope).toString()
'10 kg'
> math.eval("v = 10m/s", scope).toString()
'100 kg / s' // <-- expected this to be 10 m/s
```

I guess this a weakness with `eval` because:
1. I can do `math.unit(10, 'm/s')`.
2. It is hard to distinguish a valid multiplication vs number + unit.

To give it some context, I wanted to do some physics calculations in a [web REPL I made](https://arve0.github.io/webcas/):

![image](https://cloud.githubusercontent.com/assets/4810521/26298971/6868c8a6-3ed9-11e7-9911-4ba9310a344a.png)

Of course I could call the mass `mass` instead of `m`, but as this is quite a common way to name symbols in physics I guess others might fall in the same trap.

Any ideas? Could the "unit-syntax" be in some other way? For example:

```
1m/s  // no space -> prefer m/s as unit, do not multiply/divide if m/s in scope
1 m/s  // space -> prefer multiplication if m/s are in scope
```

**Comment by josdejong** [comment]
Thanks for bringing this up. Right now the expression parser doesn't know about units, for the parser everything is just a symbol that will be resolved at evaluation time by first looking in the `scope`, then in `math`, then whether it's a unit.

Your suggestion of distinguishing between variables and units isn't something obvious to implement. I don't think it's possible to _sometimes_ interpret a symbol `x` in `1.23 x` as a unit and _other_ times as a variable. Distinguishing with or without space would be very confusing I'm afraid.

One option would be to give an error or when overriding a unit or built-in constant or function (like `m`, `pi`, or `sqrt`), and implement some way to override this restriction.

Related: #792

**Comment by arve0** [comment]
> Distinguishing with or without space would be very confusing I'm afraid.

Yes, it would. What about some other syntax to force evaluation as unit first? A key word? Wrapped in parentheses?

> One option would be to give an error or when overriding a unit or built-in constant or function (like m, pi, or sqrt), and implement some way to override this restriction.

Not sure how I feel loosing some of the conventional symbols of physics. In one way, `F = m g` is short and concise. But on the other side, I've seen students mangle up the symbols (and I even do it my self, time to time).

Just struck me that (after writing the above), I could force units by using it's full name:
```js
$ node bin/repl.js
> scope = {}
{}
> math.eval("m = 10 kg", scope).toString()
'10 kg'
> math.eval("v = 10 meter / s", scope).toString()
'10 meter / s' // <-- as expected
```

In light of the last example, a warning would probably suffice. Is this something you would want implemented? `throw` if symbol is already defined as a unit?

**Comment by josdejong** [comment]
> Is this something you would want implemented? throw if symbol is already defined as a unit?

Yes, having an option to disable overriding existing symbols/functions and throw an error if you try makes sense. It has bitten me too a couple of times :)
